src/sundials/sundials_stepper.c

```diff
@@ -0,0 +1,214 @@
+
+
+#include <stdlib.h>
+#include <string.h>
+#include <sundials/priv/sundials_errors_impl.h>
+#include <sundials/sundials_core.h>
+
+#include "sundials/sundials_errors.h"
+#include "sundials/sundials_nvector.h"
+#include "sundials_stepper_impl.h"
+
+SUNErrCode SUNStepper_Create(SUNContext sunctx, SUNStepper* stepper)
+{
+  SUNFunctionBegin(sunctx);
+
+  *stepper = NULL;
+  *stepper = (SUNStepper)malloc(sizeof(**stepper));
+  SUNAssert(stepper, SUN_ERR_MALLOC_FAIL);
+
+  memset(*stepper, 0, sizeof(**stepper));
+
+  (*stepper)->ops = (SUNStepper_Ops)malloc(sizeof(*((*stepper)->ops)));
+  SUNAssert((*stepper)->ops, SUN_ERR_MALLOC_FAIL);
+
+  memset((*stepper)->ops, 0, sizeof(*((*stepper)->ops)));
+
+  /* initialize stepper data */
+  (*stepper)->last_flag = SUN_SUCCESS;
+  (*stepper)->sunctx    = sunctx;
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNStepper_Destroy(SUNStepper* stepper_ptr)
+{
+  SUNFunctionBegin((*stepper_ptr)->sunctx);
+
+  SUNStepper stepper = *stepper_ptr;
+
+  /* free the inner forcing and fused op workspace vector */
+  if (stepper->forcing)
+  {
+    N_VDestroyVectorArray(stepper->forcing, stepper->nforcing);
+    N_VDestroyVectorArray(stepper->fused_vectors, stepper->nforcing);
+    free(stepper->fused_scalars);
+  }
+
+  free(stepper->ops);
+  free(stepper);
+  *stepper_ptr = NULL;
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNStepper_Advance(SUNStepper stepper, sunrealtype t0,
```

Comment:
If `TryStep` advances from the last time reached and if we add a `SetStopTime` function, there could be a default implementation of `Advance` that calls `TryStep`